### PR TITLE
If we have no user crash handler we should generate additional info.

### PR DESCRIPTION
**Pull request by Penguinwizzard**
The previous way we were doing things, we'd only generate the additional
backtrace information about arguments and the like when an exception was
_totally_ uncaught. This interfered with our implicit autocatchhandlers,
which were added for some instructions. This caused divergences with the
nonative version of the debug output, where we wouldn't add these, which
caused us to only generate the additional backtrace information in a few
of the test configurations, but not all of them. By allowing creation of
the additional backtrace information when we have a catch handler but no
user catch handler, we handle these cases more uniformly.

This does, however, mean that our information on whether or not there is
a user catch handler; the existing code does this to some degree, but it
is polluted by a bunch of implied catch handlers in both interpreter and
in helper code. I added an argument to autocatchhandler to specify if we
can mark a particular instance of the autocatchhandler as user-made; the
values that I've chosen for it (i.e. where it is true) should be checked
and verified. Note that if the argument is true, that only means that it
_can_ be a user catch handler; if it's not in user code (e.g. in chakra-
generated JS) then it still won't be marked as a user catch handler.

### CODE REVIEW

File: lib/Runtime/Language/JavascriptExceptionOperators.cpp [hunk 1]
```diff
@@ -13,22 +13,30 @@ extern "C" PVOID __guard_check_icall_fptr;
...
         // If the outer try catch was already in the user code, no need to go any further.
         if (!m_previousCatchHandlerToUserCodeStatus)
         {
-            Js::JavascriptFunction* caller;
-            if (JavascriptStackWalker::GetCaller(&caller, scriptContext))
+            if (!isUserExceptionHandling)
             {
-                Js::FunctionBody *funcBody = NULL;
-                if (caller != NULL && (funcBody = caller->GetFunctionBody()) != NULL)
+                fFound = true;
```
Comment (agarwal-sandeep) [inline lib/Runtime/Language/JavascriptExceptionOperators.cpp]: >fFound = true; [](start = 16, length = 14)

nit: We actually don't need fFound and can just return.

File: lib/Runtime/Language/JavascriptExceptionOperators.cpp [hunk 2]
```diff
@@ -62,7 +67,11 @@ namespace Js
 
     bool JavascriptExceptionOperators::CrawlStackForWER(Js::ScriptContext& scriptContext)
     {
-        return Js::Configuration::Global.flags.WERExceptionSupport && !scriptContext.GetThreadContext()->HasCatchHandler();
+        // If we have exception handling, but no user exception handling, we should make
+        // the additional backtrace information available, as that increases consistency
+        // between native and nonative approaches (native code has some automatic error-
+        // handling structures that get added)
```
Comment (agarwal-sandeep) [inline lib/Runtime/Language/JavascriptExceptionOperators.cpp]: Can we add a test? None of our baselines gets effected?

File: lib/Runtime/Language/JavascriptExceptionOperators.h [hunk 3]
```diff
@@ -36,10 +36,10 @@ namespace Js
             bool m_previousCatchHandlerExists;
             bool m_previousCatchHandlerToUserCodeStatus;
             ThreadContext* m_threadContext;
-            void FetchNonUserCodeStatus(ScriptContext *scriptContext);
+            void FetchNonUserCodeStatus(ScriptContext *scriptContext, bool isUserExecptionHandling);
```
Comment (digitalinfinity) [inline lib/Runtime/Language/JavascriptExceptionOperators.h]: Nit: spelling

File: lib/Runtime/Language/JavascriptExceptionOperators.cpp (hunk 2 above)
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptExceptionOperators.cpp]: I've added a test in the new commit. The reason that none of our baselines were affected is that they pretty much all catch their exceptions in user code; this bug relates to errors _not_ caught in user code.

File: lib/Runtime/Language/JavascriptExceptionOperators.h (hunk 3 above)
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptExceptionOperators.h]: Fixed in new iteration, thanks.

File: lib/Runtime/Language/JavascriptExceptionOperators.cpp (hunk 1 above)
Comment (Penguinwizzard) [inline lib/Runtime/Language/JavascriptExceptionOperators.cpp]: Thanks, I've removed all instances of fFound and used returns to replace the related logic in the new commit on the PR.

**Comment by Penguinwizzard** [comment]
@dotnet-bot test Ubuntu ubuntu_linux_release_static please
